### ISSUE TITLE
Stringified booleans not being cast as booleans in rel_where

### ISSUE THREAD

**Issue by jmdeldin** [neo4jrb/neo4j#951]
Following up from Gitter, I'm finding boolean values (when passed in as strings, e.g., `thing.rel_where(other_thing: 'false')`) are not cast to bools in `rel_where`. From [this example](https://gist.github.com/jmdeldin/2d67e63e84021426d669): 

```ruby
# 1. this works
rs = a1.books.rel_where(enjoyed_writing: false).to_a
puts "books: #{rs.length} (should be 1)"

# 2. this does not cast 'false' into false
rs = a1.books.rel_where(enjoyed_writing: 'false').to_a
puts "books: #{rs.length} (should be 1)"
```
I worked around this in my own code, but it _seems_ like this should work.

**Comment by subvertallchris** [comment]
Good catch, I can find and fix this when I return in a few days.

**Comment by subvertallchris** [comment]
`rel_where` doesn't do any typecasting, actually. It's going to take a little work to fix it, I'm not sure it's going to happen very quickly. I also discovered that QueryProxy's typecasting only works with those converters defined within this gem, not those provided by ActiveAttr. I have some ideas for fixing it but I'm not sure I like any of them. Will post some updates when I have them but in the meantime, if you or anyone wants to take a stab at correcting this, PRs are welcome.

**Comment by jmdeldin** [comment]
Ah, thanks for the update! I'm not sure if I'll have anytime soon, but if I were to, would I start somewhere in the [Query::QueryProxy::Link](https://github.com/neo4jrb/neo4j/blob/b77ae9d24853705eaa437302bbdf8128a0dfd784/lib/neo4j/active_node/query/query_proxy_link.rb) class?

**Comment by subvertallchris** [comment]
Sort of. I think I actually managed to get it working, kind of, but here's some background:

Working backwards, the ultimate goal is `for_rel_where_clause` needs to get the ActiveRel model as an argument: https://github.com/neo4jrb/neo4j/blob/b77ae9d24853705eaa437302bbdf8128a0dfd784/lib/neo4j/active_node/query/query_proxy_link.rb#L75. If that happens, you can use https://github.com/neo4jrb/neo4j/blob/b77ae9d24853705eaa437302bbdf8128a0dfd784/lib/neo4j/active_node/query/query_proxy_link.rb#L58 or a variation thereof to perform the conversion.

Stepping back, you get into that class from this line [query_proxy.rb](https://github.com/neo4jrb/neo4j/blob/b77ae9d24853705eaa437302bbdf8128a0dfd784/lib/neo4j/active_node/query/query_proxy.rb#L339). If you change the value of `@model`, it will convert, but your chaining will be all kinds of broken because it uses that value for the context of the QueryProxy that is returned.

My solution is to just give the relationship class as another argument that only gets used by `rel_where`. It leaves the existing signature intact, doesn't break any tests, just makes me feel a little gross to have a method accepting four arguments.

The muddled type conversion becomes a problem at this point. The TLDR is that if you use a property type defined in https://github.com/neo4jrb/neo4j/blob/master/lib/neo4j/shared/typecaster.rb, QueryProxy can convert; otherwise, you're relying on ActiveAttr's converters, which don't get called. That means that String, Integer, and Boolean in particular aren't converted. The solution to _that_ is to build those locally. I just finished doing that and it seems to work but I hate adding features to heavily trafficked areas of the gem with so little though, so I'm not sure how confident I am with the implementation.

**Comment by subvertallchris** [comment]
That last problem effects `where` in addition to `rel_where`, BTW. Type conversion within `where` was a somewhat recent addition and the lack of anyone flagging an issue until now suggests it hasn't caused a problem for anyone but I would like to get it solved sooner than later.

This is a little more fuel to my "Break the dependency on ActiveAttr" fire.

**Comment by subvertallchris** [comment]
https://github.com/neo4jrb/neo4j/pull/959 fixes this, it'll be part of 5.3.0 tomorrow.

**Comment by jmdeldin** [comment]
Thanks so much for fixing this! I'll give it a try soon. :cake: :+1:

**Comment by subvertallchris** [comment]
Hey @jmdeldin, just noticed that this is still open because we waited on 5.3.0 to focus on 6.0.0. I think we'll be ready to do the alpha release of that in the next few days. Did you happen to try working off the branch or last commit referenced by that PR?

**Comment by subvertallchris** [comment]
This went out in 6.0.0.alpha.1 last night, let us know if there's any trouble.

**Comment by jmdeldin** [comment]
Thanks and sorry for the delay in testing. Is there a config option I should enable? I tried a simple test case for `where` with 6.0.0.alpha3, and it seems the strings aren't being cast:

```ruby
class Car
  include Neo4j::ActiveNode

  property :electric, type: Boolean, default: false
end

Car.destroy_all

Car.create!
Car.create!(electric: true)

p Car.all.pluck(:electric)

puts "ERROR: should only be one electric=false" if Car.where(electric: 'false').count != 1
puts "ERROR: should only be one electric=true" if Car.where(electric: 'true').count != 1

require 'pry'; binding.pry
```

results in

```
6.0.0.alpha.3
[false, true]
ERROR: should only be one electric=false
ERROR: should only be one electric=true
```

**Comment by subvertallchris** [comment]
Interesting... I thought I tested that but maybe not. I'll check it out
tonight after work.

**Comment by jmdeldin** [comment]
I _think_ I have a fix in #986, although I am not positive it's the best way to do it. It works for my example above and the tests, at least. Please let me know what you think!
